### ISSUE TITLE
Label, P, R, mAP all stay 0

### ISSUE THREAD

**Issue by TimRobben** [JDSobek/MedYOLO#37]
First of all thank you for providing your model to us!
I'm trying to detect a patient has a small object in the heart with the following mean measurements in mm:
 
d_mm     22.53 
w_mm    13.72  
h_mm     7.87  
I have a small dataset consisting of around 20 positive cases and a couple 100 negative cases, which I know is very small. I tried upsampling the positive cases 4x and making the training set more balanced that way. I am considering some heavy augmentation to improve the model. 

I'm facing an issue where the model is not making any predictions that reach the thresh hold in the validation, even if I put the confidence (of both training and validation) to 0.01. One of my runs reached an early stopping at 367 epochs, and another kept running until the 1000, both with all metrics staying 0. I even tried overfitting on one positive example to see if it will detect this, but so far I haven't gotten any results (it is still running, at 100 epochs currently). 

In your expertise, do you think it is even possible for the model to learn such a small object with this little data? Or do you have any other advice of things I could look into? 

I look forward to your response, and if you need any more information I can provide that.

**Comment by JDSobek** [comment]
I don't know the relative sizes, but a colleague [used it for aortic root localization](https://www.jacc.org/doi/10.1016/j.jacc.2025.03.145) with good success. I don't remember her dataset being very large but it was all positive cases.

One of the datasets I used was small (80) and the model had trouble detecting a couple structures that weren't in every example (prostate and uterus) with similar prevalence to your dataset IIRC. I think the model, at least with the relatively simple implementation used for the data pipeline, struggles with detecting rare objects. That might explain your results as much as the object's small size.

One example probably won't let the model update enough times to perform well, but maybe running all of your positive cases with a relatively small batch size and some more epochs (maybe 2-4 and 2000 respectively) will give you a better idea of whether the model can detect the object at all or if it's too small.

If you're primarily focused just on detecting somewhat rare objects and localizing them isn't important, you may want to use another model. If you need to do both you may have better luck using a two-stage approach, detecting positive cases with another model and feeding the positive cases into this, or another, localization model.

You might want to look into MONAI, which I think has some 3D models with pretrained weights that you could transfer learn to your task.

**Comment by TimRobben** [comment]
Thank you for your quick response! 
I'm aiming to localize the artifact in the heart. Unfortunately the object is smaller than your collegues research, about 56 × 34 × 20 voxels (depth × width × height) in a CT of 512x512x350. 
I have tried what you recommended and ran the small model with my 20 positive cases with a batch size of 2 for 2000 epochs. The goal was to overfit so I put the same 20 cases in both the train and validation set. After 947 it didn't improve any more, so it stopped. The results are:
Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95:
    all         21         21      0.024     0.0476    0.00383   0.000767

So that's not great, and I don't expect any magic fix, but I was wondering if you might have any recommendation as to any improvements. I know you said that in your experience, if the small model cannot detect it, the larger model will often not do any better, but is this also the case for small objects? 
Thank you again for your help!

**Comment by JDSobek** [comment]
For me that's what happened with small objects, but if you have the time and spare computational power you could always test it. That was mostly in reference to LIDC and some weird things I noticed with BRATS, which I spent a lot less time working on than the heart dataset and don't know as well.

Off the top of my head here are a few other things that might be worth trying.

1. MedYOLO should do well detecting the heart (mind you I annotated like 600 cases...), so you could try annotating your hearts, training a model to detect them, and then training for artifact detection on zoomed in views of the heart coming from the first model. For testing purposes you could skip training the heart detection model and try training MedYOLO to detect artifacts when you pass it a NIfTI containing just the volume around the heart, then if that works going back to train heart detection.
2. If you haven't already, you can try training 2-D YOLO models by splitting your NIfTIs into png/jpg files. Those allow for heavier augmentation that might let you run for longer before early stopping kicks in. They also have lighter VRAM requirements which means you can zoom the 2D imaging to bigger sizes (where MedYOLO smushes everything into 350x350x350 cubes and might be shrinking useful features) or avoid modifying the aspect ratios like MedYOLO unfortunately has to do.
3. On that note you can try making the cubes bigger. This is very VRAM intensive and didn't help me with LIDC, but might help with less diffuse object detection, especially if you're already running with low batch size for other reasons and have spare VRAM. 350 is basically the smallest cube that can run through the model without causing issues from downsampling.
4. You could try turning off early stopping (set it to 0) or giving it a much longer patience (Maybe 500 for 2000 epoch training). Especially when training for more epochs, in one-cycle training there's a lengthy period where the learning rate is really high and the model should achieve its best performance during the descent from that period, so forcing the model to get further past that part before it can stop may give better results. There's no promises that will give great results but if it even starts to see the artifacts it's a sign that the dataset may be limiting you as much or more than the model is. The model should be saving the best weights either way so unless you're strapped for GPU time there's not really any downsides to trying this.
5. If it's at all possible to get more data, that's probably your best bet for getting any method working. I needed 60 training cases (with ~100% true positive object presence) with 5x augmentation for the smallest dataset to show decent results, and for heart detection without heavy augmentation there was a substantial improvement going from ~400 training cases to ~600.

5 is your best bet but the hardest to do, 3 and 4 are the easiest to try (alongside trying a larger model) since there are just a couple settings to change when you run `train.py`. 1 and 2 only need a little extra data annotation to train the model(s) but require extra code to go from trained model(s) to inference pipeline.

**Comment by TimRobben** [comment]
When you say you used 5X augmentation, do you mean you augmented the data before training or did you use the augmentations you implemented in medyolo? And what augmentation did you use?

**Comment by JDSobek** [comment]
The Data section on the paper (see the Arxiv link in the readme) gives the details exactly, but for my smallest dataset I did rotation augmentation on the training examples 5 times, saved those rotated examples, calculated their bounding boxes, and used those augmented examples alongside the original examples to train the model. So in the end I went from 60 -> 360 training examples.

Rotation augmentation isn't implemented in MedYOLO because I couldn't figure out how to translate the openCV2 matrix operations YOLOv5 uses to transform bounding boxes into some NIfTI compatible 3-D operations, and I wasn't certain enough about what exactly was happening to the bounding boxes to do it in torch.

**Comment by TimRobben** [comment]
Thank you again for all your help! A little update; I tried upsampling x7 to check if anything would happen and luckily it started to learn! The other thing I changed was using lower thresholds, so i edited this line  orig: iouv = torch.linspace(0.5, 0.95, 10).to(device)  # iou vector for mAP@0.5:0.95 to   iouv = torch.linspace(0.1, 0.5, 9).to(device)  # iou vector for mAP@0.1:0.50. 
Currently, I augmented my data 7 times, but ran into an empty batches problem, since I have 50% negative images (without label). I am using image-weights, but its not enough unfortunately. Now I'm working on making a custom sampler, so that there is at least 1 positive case in each batch. So currently I edited nifti_dataloader() to use this as the sampler:

````
class SafeBatchSampler(Sampler):
    """
    Works like the default RandomSampler+BatchSampler, except it
    guarantees every batch contains ≥1 positive (len(label)>0).

    • Uses a full random permutation of the dataset each epoch
      (so class statistics over the epoch are unchanged).
    • Only when a batch happens to have zero PFO cases does it
      replace one item with a still‑unused positive index.
    """

    def __init__(self, dataset, batch_size, drop_last=True):
        self.dataset     = dataset
        self.batch_size  = batch_size
        self.drop_last   = drop_last
        # build the mask once
        self.pos_idx = [i for i,l in enumerate(dataset.labels) if len(l)]
        self.neg_idx = [i for i,l in enumerate(dataset.labels) if not len(l)]
        if not self.pos_idx:
            raise RuntimeError("SafeBatchSampler: dataset has no positives!")
        self.gen = torch.Generator()

    def __iter__(self):
        # full random order each epoch
        order = torch.randperm(len(self.dataset), generator=self.gen).tolist()
        # pointer into unused positives for “swap‑in”
        pos_pool = self.pos_idx.copy()
        torch.random.manual_seed(torch.randint(0, 2**31-1, (1,)).item())
        random.shuffle(pos_pool)        # shuffle once per epoch
        pos_ptr = 0

        for i in range(0, len(order), self.batch_size):
            batch = order[i : i + self.batch_size]
            if len(batch) < self.batch_size and self.drop_last:
                break                         # mimic default behaviour
            # ---- guarantee ≥1 positive ---------------------------------
            if all(len(self.dataset.labels[j]) == 0 for j in batch):
                # swap first element with an unused positive
                if pos_ptr >= len(pos_pool):          # recycle if needed
                    pos_pool = self.pos_idx.copy()
                    random.shuffle(pos_pool)
                    pos_ptr = 0
                batch[0] = pos_pool[pos_ptr]
                pos_ptr += 1
            yield batch

    def __len__(self):
        n = len(self.dataset) // self.batch_size
        return n if self.drop_last else n + (len(self.dataset) % self.batch_size > 0)
```

As you have a much better understanding of everything, do you see any issues with this?

**Comment by JDSobek** [comment]
Changing `iouv` in `val.py` won't change what the model is doing, just what's reported to the screen during validation, so it shouldn't cause any problems. IIRC [0.1, 0.5] mAP is what nnDetect uses, so that range isn't unprecedented either.

I haven't really used custom samplers before. I think it looks fine, but a result from the torch forums or stack overflow would probably be more reliable.
